File: server/skills/allow_folder.js
// Skill: grants Jarvis permission to read/write/move files inside one
// specific folder. This is the ONLY way the files connector's allowlist
// ever grows — there is no settings screen for it (see CLAUDE.md's "App
// Control connectors" section). A file tool hitting a folder outside the
// allowlist fails with a plain error; the model is expected to ask the user
// in conversation and only then call this. `confirm: 'always'` reuses the
// existing read-back-and-confirm gate (skills/index.js) rather than trusting
// the model's own judgment about what counts as a real "yes."
//
// Meta skill (excluded from the scheduled-task/briefing picker and from any
// unattended/background turn) — granting new folder access only ever makes
// sense as part of a live, two-way conversation, never something a scheduled
// prompt-mode task should be able to trigger on its own.

import path from 'node:path';
import { getOrCreateSingleton, updateConnector } from '../connectors/store.js';

// A small, non-editable floor under the folder allowlist — the same
// "sensible hardcoded defaults" philosophy as control/safety.js's window/
// process blocklist, applied to the filesystem: these can never be granted,
// no matter what's asked for, since they're core to Windows itself.
const DENYLISTED_ROOTS = ['C:\\Windows', 'C:\\Program Files', 'C:\\Program Files (x86)'];

function isDenylisted(resolved) {
  const lower = resolved.toLowerCase();
  return DENYLISTED_ROOTS.some((root) => {
    const rootLower = root.toLowerCase();
    return lower === rootLower || lower.startsWith(rootLower + path.sep);
  });
}

export default {
  name: 'allow_folder',
  meta: true,
  description:
    'Grants Jarvis permission to read, write, and move files inside one specific folder. Only call this ' +
    'after a file operation failed because the folder was not allowed yet, and the user has agreed to let ' +
    'Jarvis use that folder.',
  confirm: 'always',
  parameters: {
    type: 'object',
    properties: {
      folder: { type: 'string', description: 'The absolute folder path to allow, e.g. C:\\Users\\you\\Documents.' },
    },
    required: ['folder'],
  },
  summarize(args) {
    return `Allow Jarvis to read, write, and move files inside "${args.folder}"?`;
  },
  async run(args) {
    const resolved = path.resolve(args.folder);
    if (isDenylisted(resolved)) {
      return { ok: false, error: `"${resolved}" is a core Windows folder — Jarvis can't be granted access to it.` };
    }
    const connector = getOrCreateSingleton('files', { label: 'Files' });
    const allowedFolders = connector.config?.allowedFolders || [];
    if (allowedFolders.includes(resolved)) {
      return { ok: true, alreadyAllowed: resolved };
    }
    updateConnector(connector.id, { config: { allowedFolders: [...allowedFolders, resolved] } });
    return { ok: true, granted: resolved };
  },
};

File: backend/jarvis/tools/allow_folder.py
"""Remembering that the user said yes to a folder.

The file allowlist starts empty and grows only through conversation: an
operation outside it fails with a plain error, the model asks, and this records
the answer. That shape is deliberate — there is no settings screen where someone
grants blanket access to their disk in advance and forgets they did.

Confirmed like anything else that changes what Jarvis may reach, and read back
with the real path, because "can I have access to that folder?" and "can I have
access to C:\\Users" are very different questions.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from ..capabilities import CapabilitySpec, Risk


def _run(registry: Any, path: str = "") -> dict[str, Any]:
    from ..connectors import store
    from ..connectors.capabilities import sync

    target = Path(str(path or "").strip()).expanduser()
    if not str(path or "").strip():
        return {"ok": False, "error": "I didn't catch which folder."}
    if not target.exists():
        return {"ok": False, "error": f'"{path}" isn\'t a folder on this computer.'}
    if not target.is_dir():
        return {"ok": False, "error": f'"{path}" is a file, not a folder.'}

    connector = store.get_or_create_singleton("files")
    config = dict(connector.get("config") or {})
    folders = list(config.get("allowedFolders") or [])
    resolved = str(target.resolve())
    if resolved in folders:
        return {"ok": True, "path": resolved, "alreadyAllowed": True,
                "note": "That folder was already allowed."}

    folders.append(resolved)
    config["allowedFolders"] = folders
    store.update_connector(connector["id"], {"config": config})

    # The file tools only exist once there is somewhere to use them.
    sync(registry)
    return {"ok": True, "path": resolved,
            "note": "Files in that folder and below it can be read and written from now on."}


def build(registry: Any) -> list[CapabilitySpec]:
    """Built with the registry passed in: a tool must never reach for the shared
    one, and this needs it to make the file tools appear."""
    return [CapabilitySpec(
        id="builtin.allow_folder",
        name="allow_folder",
        description=("Remember that the user has given permission to work with files in "
                     "one folder. Only call it after actually asking them about that "
                     "specific folder and getting a yes — never to pre-empt a refusal, "
                     "and never on a folder they didn't name."),
        input_schema={"type": "object", "properties": {
            "path": {"type": "string",
                     "description": "The folder's full path, exactly as they gave it."}},
            "required": ["path"]},
        risk=Risk.MEDIUM,
        handler=lambda **args: _run(registry, **args),
        timeout_s=10.0,
        summarize=lambda args: (f'Allow reading and writing files in "{args.get("path")}" '
                                f"and everything below it?"),
    )]
